eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/MessageOnion.scala

```diff
@@ -83,31 +83,56 @@ object MessageOnion {
       if (records.get[InvoiceRequest].nonEmpty) return Left(ForbiddenTlv(UInt64(64)))
       if (records.get[Invoice].nonEmpty) return Left(ForbiddenTlv(UInt64(66)))
       if (records.get[InvoiceError].nonEmpty) return Left(ForbiddenTlv(UInt64(68)))
+      if (records.unknown.nonEmpty) return Left(ForbiddenTlv(records.unknown.head.tag))
       BlindedRouteData.validateMessageRelayData(blindedRecords).map(blindedRecords => IntermediatePayload(records, blindedRecords, nextBlinding))
     }
   }
 
   /** Per-hop payload for a final node. */
-  case class FinalPayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]) extends PerHopPayload {
+  sealed trait FinalPayload extends PerHopPayload {
+    val blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]
     val pathId_opt: Option[ByteVector] = blindedRecords.get[RouteBlindingEncryptedDataTlv.PathId].map(_.data)
```

Comment:
nit: `val` in `trait`s can create instantiation issues at runtime, we've been bitten by that in the past

```suggestion
    def blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]
    def pathId_opt: Option[ByteVector] = blindedRecords.get[RouteBlindingEncryptedDataTlv.PathId].map(_.data)
```

Reply:
Done